### PR TITLE
Bug 1905095: Verify layer MIME types in images

### PR DESCRIPTION
**Pull request by adambkaplan**
Verify that the pushed image has the gzip-ed layer MIME type for
docker v2schema2.

### CODE REVIEW

**Comment by adambkaplan** [comment]
/hold

This test should fail until containers/image in openshift/builder is bumped to v5.10.0 or higher.

File: go.mod [hunk 1]
```diff
@@ -8,6 +8,7 @@ require (
 	github.com/apparentlymart/go-cidr v1.1.0
 	github.com/auth0/go-jwt-middleware v0.0.0-20201030150249-d783b5c46b39 // indirect
 	github.com/boltdb/bolt v1.3.1 // indirect
+	github.com/containers/image/v5 v5.9.0
```
Comment (gabemontero) [inline go.mod:11]: they just cut v5.10.0 per our slack exchange

**Comment by adambkaplan** [comment]
boo transitive deps - c/image moved us from v0 to v1 libraries for a few things (opencontainers one of them)

File: test/extended/builds/digest.go [hunk 2]
```diff
@@ -75,8 +77,15 @@ func testBuildDigest(oc *exutil.CLI, buildFixture string, buildLogLevel uint) {
...
-		ist, err := oc.ImageClient().ImageV1().ImageStreamTags(oc.Namespace()).Get(context.Background(), "test:latest", v1.GetOptions{})
+		ist, err := oc.ImageClient().ImageV1().ImageStreamTags(oc.Namespace()).Get(context.Background(), "test:latest", metav1.GetOptions{})
 		o.Expect(err).NotTo(o.HaveOccurred())
 		o.Expect(br.Build.Status.Output.To.ImageDigest).To(o.Equal(ist.Image.Name))
+
+		g.By("checking that the image layers have valid docker v2schema2 MIME types")
+		image, err := oc.ImageClient().ImageV1().Images().Get(context.Background(), br.Build.Status.Output.To.ImageDigest, metav1.GetOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+		for _, layer := range image.DockerImageLayers {
+			o.Expect(layer.MediaType).To(o.Equal(manifestschema2.MediaTypeLayer))
```
Comment (adambkaplan) [inline test/extended/builds/digest.go:88]: @nalind per our discussion in Slack, we are requiring all pushed layers to have the default v2s2 MIME type (`application/vnd.docker.image.rootfs.diff.tar.gzip`)

Comment (nalind) [reply test/extended/builds/digest.go:88]: Yep, if we know it's v2s2, I think that's correct.  Given that we generated it, I don't think we need to consider MediaTypeForeignLayer.

**Comment by adambkaplan** [comment]
@nalind we are still seeing the bad layers in the resulting image:

```
fail [github.com/openshift/origin/test/extended/builds/digest.go:88]: Expected
    <string>: application/vnd.docker.image.rootfs.diff.tar
to equal
    <string>: application/vnd.docker.image.rootfs.diff.tar.gzip
```

**Comment by adambkaplan** [comment]
@nalind potentially relevant log from the latest failure - this layer had the uncompressed MIME type in the resulting image as reported in the Image object:

```
Getting image source signatures
time="2021-02-02T15:30:10Z" level=debug msg="Manifest has MIME type application/vnd.docker.distribution.manifest.v2+json, ordered candidate list [application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.v1+prettyjws, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.docker.distribution.manifest.v1+json]"
time="2021-02-02T15:30:10Z" level=debug msg="... will first try using the original manifest unmodified"
time="2021-02-02T15:30:10Z" level=debug msg="suggesting cached blob with digest \"sha256:b8273964fced45310fd2d348e3adeb9ad5e8f5ecc8f3de5b43b176dabd963110\" and compression 2 in place of blob with digest \"sha256:683f499823be212bf04cb9540407d8353803c25d0d9eb5f2fdb62786d8b95ead\""
...
time="2021-02-02T15:30:10Z" level=debug msg="Checking /v2/e2e-test-build-sti-labels-dkfp4/test/blobs/sha256:b8273964fced45310fd2d348e3adeb9ad5e8f5ecc8f3de5b43b176dabd963110"
...
time="2021-02-02T15:30:10Z" level=debug msg="HEAD https://image-registry.openshift-image-registry.svc:5000/v2/e2e-test-build-sti-labels-dkfp4/test/blobs/sha256:b8273964fced45310fd2d348e3adeb9ad5e8f5ecc8f3de5b43b176dabd963110"
...
Copying blob sha256:b8273964fced45310fd2d348e3adeb9ad5e8f5ecc8f3de5b43b176dabd963110
time="2021-02-02T15:30:11Z" level=debug msg="Detected compression format gzip"
time="2021-02-02T15:30:11Z" level=debug msg="Using original blob without modification"
time="2021-02-02T15:30:11Z" level=debug msg="Checking /v2/e2e-test-build-sti-labels-dkfp4/test/blobs/sha256:b8273964fced45310fd2d348e3adeb9ad5e8f5ecc8f3de5b43b176dabd963110"
time="2021-02-02T15:30:11Z" level=debug msg="HEAD https://image-registry.openshift-image-registry.svc:5000/v2/e2e-test-build-sti-labels-dkfp4/test/blobs/sha256:b8273964fced45310fd2d348e3adeb9ad5e8f5ecc8f3de5b43b176dabd963110"
time="2021-02-02T15:30:11Z" level=debug msg="... not present"
...
time="2021-02-02T15:30:11Z" level=debug msg="PUT https://image-registry.openshift-image-registry.svc:5000/v2/e2e-test-build-sti-labels-dkfp4/test/blobs/uploads/0fc802eb-93a3-4dba-a19b-bd74f56f9977?_state=EXqlxU5WauvCPRqm_6A4urdZM-SzaHNhWU4uzl15PWN7Ik5hbWUiOiJlMmUtdGVzdC1idWlsZC1zdGktbGFiZWxzLWRrZnA0L3Rlc3QiLCJVVUlEIjoiMGZjODAyZWItOTNhMy00ZGJhLWExOWItYmQ3NGY1NmY5OTc3IiwiT2Zmc2V0Ijo3NzIwNjIsIlN0YXJ0ZWRBdCI6IjIwMjEtMDItMDJUMTU6MzA6MTFaIn0%3D&digest=sha256%3Ab8273964fced45310fd2d348e3adeb9ad5e8f5ecc8f3de5b43b176dabd963110"
...
time="2021-02-02T15:30:12Z" level=debug msg="Upload of layer sha256:b8273964fced45310fd2d348e3adeb9ad5e8f5ecc8f3de5b43b176dabd963110 complete"
...
```

**Comment by nalind** [comment]
That "suggesting cached blob" message is coming from the blob cache (which is a different thing than the blob info cache).  It hasn't been updated in some time, and it seems to be undermining the blob info cache's efforts here.

**Comment by gabemontero** [comment]
@adambkaplan assuming we get a passing e2e-gcp-builds, are you going to want to leave that debug commit in there, or do you want to remove it before we lgtm it?

**Comment by adambkaplan** [comment]
@gabemontero I'll squash the commits later tonight, want to see a clean run of the builds suite first.

**Comment by adambkaplan** [comment]
GCP BUILDS PASSED!!! :tada: :tada: :tada:

**Comment by LorbusChris** [comment]
@bparees could you override the permafailing upgrade test again?

**Pull request merged**